nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server.hpp

```diff
@@ -0,0 +1,374 @@
+  {
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  // Derived classes can override any of the following methods to hook into the
+  // processing for the action: on_goal_received, on_loop, and is_cancelling
+
+  // Could check if goal is valid and put values
+  // on the blackboard which depend on the received goal
+  virtual bool on_goal_received() = 0;
+
+  // Could define execution that happens on one iteration through the BT
+  // Preempt current goal when new goal is received
+  // Publish feedback
+  virtual void on_loop() = 0;
+
+  // Decides if cancel has been requested for the current goal
+  virtual bool is_canceling() = 0;
+
+protected:
+  /**
+   * @brief Configures member variables
+   *
+   * Initializes action server for, builds behavior tree from xml file,
+   * and calls user-defined onConfigure.
+   * @param state Reference to LifeCycle node state
+   * @return SUCCESS or FAILURE
+   */
+  nav2_util::CallbackReturn on_configure(const rclcpp_lifecycle::State & /*state*/) override
+  {
+    RCLCPP_INFO(get_logger(), "Configuring");
+
+    // use suffix '_rclcpp_node' to keep parameter file consistency #1773
+    auto options = rclcpp::NodeOptions().arguments(
+      {"--ros-args",
+        "-r", std::string("__node:=") + get_name() + "_rclcpp_node",
+        "--"});
+    // Support for handling the topic-based goal pose from rviz
+    client_node_ = std::make_shared<rclcpp::Node>("_", options);
+
+    tf_ = std::make_shared<tf2_ros::Buffer>(get_clock());
+    auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+      get_node_base_interface(), get_node_timers_interface());
+    tf_->setCreateTimerInterface(timer_interface);
+    tf_->setUsingDedicatedThread(true);
+    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_, this, false);
+
+    action_server_ = std::make_unique<ActionServer>(
+      get_node_base_interface(),
+      get_node_clock_interface(),
+      get_node_logging_interface(),
+      get_node_waitables_interface(),
+      action_name_, std::bind(&BtActionServer<ActionT>::executeCallback, this));
+
+    // Get the libraries to pull plugins from
+    plugin_lib_names_ = get_parameter("plugin_lib_names").as_string_array();
+
+    // Create the class that registers our custom nodes and executes the BT
+    bt_ = std::make_unique<nav2_behavior_tree::BehaviorTreeEngine>(plugin_lib_names_);
+
+    // Create the blackboard that will be shared by all of the nodes in the tree
+    blackboard_ = BT::Blackboard::create();
+
+    // Put items on the blackboard
+    blackboard_->set<rclcpp::Node::SharedPtr>("node", client_node_);  // NOLINT
+    blackboard_->set<std::shared_ptr<tf2_ros::Buffer>>("tf_buffer", tf_);  // NOLINT
+    blackboard_->set<std::chrono::milliseconds>("server_timeout", std::chrono::milliseconds(10));  // NOLINT
+    blackboard_->set<bool>("initial_pose_received", false);  // NOLINT
+    blackboard_->set<int>("number_recoveries", 0);  // NOLINT
```

Comment:
The problems with some of this is that this doesn't apply to all uses of the BT action server. For BT Navigator where it's handling the navigation process, these all make sense. But for a behavior tree based autonomy application calling `NavigateToPose`, many of these aren't used since its all higher level logic

Reply:
Though you have some of those specific ones in the BT navigator, so I think they can just be removed here.

Reply:
> Though you have some of those specific ones in the BT navigator, so I think they can just be removed here.

Yup, I was supposed to remove these but ended up forgetting. Though now that I noticed, `initial_pose_received` is not being used anywhere in navigation2 other than the `nav2_behavior_tree::initialPoseReceived` BT node which itself is not used anywhere. I think `initial_pose_received` and `nav2_behavior_tree::initialPoseReceived` can be removed altogether.